File: problems/problemset/ttl-cache/C++/runner/Testbench.cpp
#include <bits/stdc++.h>
#include <nlohmann/json.hpp>
#include <chrono>
#include <sys/resource.h>
#include "../code/TTLCache.h"

using json = nlohmann::json;

static long peakMemoryKB() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss;
}

int main() {
    std::string testcase_str((std::istreambuf_iterator<char>(std::cin)),
                              std::istreambuf_iterator<char>());

    json testcase = json::parse(testcase_str);

    auto input = testcase["input"].get<std::string>();
    std::stringstream ss(input);
    std::string cmd;

    std::string output;
    TTLCache cache;

    auto start = std::chrono::high_resolution_clock::now();

    while (ss >> cmd) {
        if (cmd == "SET") {
            std::string key, value;
            long long ttl;
            ss >> key >> value >> ttl;
            cache.set(key, value, ttl);
        } else if (cmd == "GET") {
            std::string key;
            ss >> key;
            std::string val = cache.get(key);
            output += (val.empty() ? "\"\"" : val) + "\n";
        } else if (cmd == "DEL") {
            std::string key;
            ss >> key;
            cache.del(key);
        } else if (cmd == "SIZE") {
            output += std::to_string(cache.size()) + "\n";
        } else if (cmd == "TICK") {
            long long delta;
            ss >> delta;
            cache.tick(delta);
        }
    }

    auto end = std::chrono::high_resolution_clock::now();
    double time_ms   = std::chrono::duration<double, std::milli>(end - start).count();
    long   memory_kb = peakMemoryKB();

    json result;
    result["actual_output"] = output;
    result["time_ms"]       = time_ms;
    result["memory_kb"]     = memory_kb;

    std::cout << "__BEGIN_RESULT__" << result.dump() << "__END_RESULT__";
}

File: problems/problemset/ttl-cache/Python/runner/testbench.py
import sys
import json
import time
import resource

sys.path.insert(0, '/workspace/code')
from ttl_cache import TTLCache

def main():
    testcase = json.loads(sys.stdin.read())
    cache = TTLCache()
    output_lines = []

    start = time.perf_counter()

    for line in testcase['input'].splitlines():
        parts = line.split()
        if not parts:
            continue
        cmd = parts[0]
        if cmd == 'SET':
            cache.set(parts[1], parts[2], int(parts[3]))
        elif cmd == 'GET':
            val = cache.get(parts[1])
            output_lines.append('""' if not val else val)
        elif cmd == 'DEL':
            cache.delete(parts[1])
        elif cmd == 'SIZE':
            output_lines.append(str(cache.size()))
        elif cmd == 'TICK':
            cache.tick(int(parts[1]))

    elapsed_ms = (time.perf_counter() - start) * 1000
    memory_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    actual_output = '\n'.join(output_lines) + '\n' if output_lines else ''

    result = {'actual_output': actual_output, 'time_ms': elapsed_ms, 'memory_kb': memory_kb}
    sys.stdout.write(f'__BEGIN_RESULT__{json.dumps(result)}__END_RESULT__')

main()
